armi/reactor/composites.py

```diff
@@ -771,66 +778,80 @@ def _deriveUndefinedVolume(self):
 
-        leftover = []
-        processed = []  # for input debugging
-        totalVolume = 0.0
-        for child in self.getChildren():
-            # so far, only coolants do this. ThRZBlock's with these must have
-            # area set specifically by input.
-            if child.__class__ is components.DerivedShape:
-                leftover.append(child)
-            else:
-                totalVolume += child.getVolume()
-                processed.append(child)
-        derivedVolume = 0.0
-        if len(leftover) == 1:
-            left = leftover[0]
-            derivedVolume = self.getMaxVolume() - totalVolume
-            if derivedVolume < 0:
-                runLog.error(
-                    "Negative remaining volume of {0} cm^3 for {1} in {2}. "
-                    "Check geometry (is pitch correct?). \nMax volume in this object is {3}\n"
-                    "Total of others is {4}\n"
-                    "Processed children include:\n{5}"
-                    "".format(
-                        derivedVolume,
-                        left,
-                        self,
-                        self.getMaxVolume(),
-                        totalVolume,
-                        "\n".join(
-                            [
-                                f"{child:40s}, {child.getArea():.5e}"
-                                for child in processed
-                            ]
-                        ),
-                    )
-                )
-                raise RuntimeError(
-                    "Negative remaining volume ({}) in {}".format(derivedVolume, self)
-                )
-            left.setVolume(derivedVolume)
-            totalVolume += derivedVolume
-        elif leftover:
-            runLog.warning(
-                "Gluttony Error incoming, total components {}"
-                "".format(self.getChildren())
-            )
-            for c in self.getChildren():
-                runLog.warning(
-                    "volume {} params {} for component {}"
-                    "".format(c.getVolume(), c.p, c.getName())
-                )
-            runLog.error(
-                "These are leftovers: {0}\n"
-                " Cannot deduce area of more than one component".format(leftover)
-            )
-            raise RuntimeError("Gluttony Error too many leftovers.")
-        return derivedVolume
+        # Determine the number of `DerivedShape` components and non-derived shape components
+        derivedShapeComps = [c for c in self.getChildren() if c.__class__ is components.DerivedShape]
+        otherComps = [c for c in self.getChildren() if c not in derivedShapeComps]
+
+        otherCompAreas = sum([c.getArea() for c in otherComps])
+        otherCompVolumes = sum([c.getVolume() for c in otherComps])
+
+        remainingArea = self.getMaxArea() - otherCompAreas
+        remainingVolume = self.getMaxVolume() - otherCompVolumes
```

Comment:
I think this is a potentially serious performance hit for blocks that do not have DerivedShapes in them, right? Can you try to keep this in a conditional that only happens when there's a derived shape?

(BTW we've been thinking of making the "DerivedShape" behavior specific to a block subclass that has that behavior so that may be the future state of this).

Reply:
So my understanding is that this only gets called on the `DerivedShape` class, but I agree that this can be put underneath the check that there is only 1 `DerivedShape` in the block.

Reply:
If this only exists to make DerivedShape possible, then this logic should live in the DerivedShape class. This whole method is implemented using public data and interfaces.

Reply:
wait.... and its _private_?? Yeah, definitely just move this whole thing over to the DerivedShape class.